### PR TITLE
pageSize for getEncodedUserOrders

### PR DESCRIPTION
**Pull request by josojo**
This address the pagination proposal from here:
https://github.com/gnosis/dex-contracts/issues/386#issuecomment-567574354

This would improve the performance of the view, if there are traders with a lot of orders

### CODE REVIEW

**Review by fleupold — CHANGES_REQUESTED** [review]
Let’s offer this as an additional method please so that we can still use the old one for now

File: contracts/BatchExchange.sol [hunk 1]
```diff
@@ -381,8 +381,8 @@ contract BatchExchange is EpochTokenLocker {
       * @param user address of user whose orders are being queried
       * @return encoded bytes representing all orders
       */
-    function getEncodedUserOrders(address user) public view returns (bytes memory elements) {
-        for (uint256 i = 0; i < orders[user].length; i++) {
+    function getEncodedUserOrders(address user, uint256 offset, uint16 pageSize) public view returns (bytes memory elements) {
```
Comment (nlordell) [inline contracts/BatchExchange.sol]: `offset + pagesize` is an `orderIndex` so shouldn't `offset` also be a `uint16`?

File: contracts/BatchExchange.sol [hunk 2]
```diff
@@ -398,7 +398,7 @@ contract BatchExchange is EpochTokenLocker {
             address user = allUsers.first();
```
Comment (nlordell) [inline contracts/BatchExchange.sol]: I think `getEncodedOrders` also needs to support pagination, or else the driver will run into the same problems when it tries to get the order book to pass to the solver.

Comment (josojo) [reply contracts/BatchExchange.sol]: I think the requirement is only needed for the front-end as the solver can just point to another fullnode/our fullnode, which does not have the same restrictions as infura

**Review by bh2smith — APPROVED** [review]
This is really cool! Should we make a test to ensure it is working as expected!?

File: contracts/BatchExchange.sol (hunk 2 above)
Comment (nlordell) [reply contracts/BatchExchange.sol]: Oh good point!

File: test/stablex/batch_exchange.js [hunk 3]
```diff
@@ -1654,6 +1654,44 @@ contract("BatchExchange", async accounts => {
...
   })
+  describe("getEncodedUserOrdersPaginated()", async () => {
+    it("returns correct orders considering offset and pageSize", async () => {
+      const batchExchange = await setupGenericStableX()
+      const zeroBN = new BN(0)
+      const tenBN = new BN(10)
+      const twentyBN = new BN(20)
+
+      const batchId = (await batchExchange.getCurrentBatchId.call()).toNumber()
+      const OrderInfo = {
```
Comment (fleupold) [inline test/stablex/batch_exchange.js]: nit: lower case

File: test/stablex/batch_exchange.js [hunk 4]
```diff
@@ -1654,6 +1654,44 @@ contract("BatchExchange", async accounts => {
...
+          OrderInfo.sellToken,
+          OrderInfo.validUntil,
+          OrderInfo.priceNumerator.addn(i),
+          OrderInfo.priceDenominator
+        )
+      }
+
+      // get 2nd order with getEncodedUserOrdersPaginated(user_1, 1, 1)
+      const expectedOrder = OrderInfo
+      expectedOrder.priceNumerator = expectedOrder.priceNumerator.addn(1)
```
Comment (fleupold) [inline test/stablex/batch_exchange.js]: ubertnit (and more high-level feedback): I find this unit test a bit logic intense. I'd just create orders with all value 0, all values 1, 2, and then assert that we get the one 1 all values 1.

File: contracts/BatchExchange.sol [hunk 5]
```diff
@@ -377,6 +377,25 @@ contract BatchExchange is EpochTokenLocker {
         return IdToAddressBiMap.hasAddress(registeredTokens, addr);
     }
 
+    /** @dev View returning all byte-encoded sell orders for specified user
+      * @param user address of user whose orders are being queried
+      * @param offset uint determining the starting orderIndex
+      * @param pageSize uint determining the count of elements to be viewed
+      * @return encoded bytes representing all orders
+      */
+    function getEncodedUserOrdersPaginated(address user, uint16 offset, uint16 pageSize)
```
Comment (fleupold) [inline contracts/BatchExchange.sol:386]: How about calling this method inside `getEncodedUserOrders` with pageSize length and offset 0?

File: test/stablex/batch_exchange.js [hunk 6]
```diff
@@ -1654,6 +1654,44 @@ contract("BatchExchange", async accounts => {
...
+      const auctionElements = decodeAuctionElements(await batchExchange.getEncodedUserOrdersPaginated(user_1, 1, 1))
+      assert.equal(JSON.stringify(auctionElements), JSON.stringify([expectedOrder]))
```
Comment (bh2smith) [inline test/stablex/batch_exchange.js]:
```suggestion
      assert.equal(JSON.stringify(auctionElements[0]), JSON.stringify(expectedOrder))
```

File: contracts/BatchExchange.sol [hunk 7]
```diff
@@ -377,6 +377,25 @@ contract BatchExchange is EpochTokenLocker {
...
+        public
+        view
+        returns (bytes memory elements)
+    {
+        for (uint256 i = offset; i < Math.min(orders[user].length, offset + pageSize); i++) {
```
Comment (fleupold) [inline contracts/BatchExchange.sol]: nit: should be u16 or u32 (if we want to support overflowing)

**Review by fleupold — APPROVED** [review]

File: test/stablex/batch_exchange.js (hunk 6 above)
Comment (bh2smith) [reply test/stablex/batch_exchange.js]: nit: I think retrieving the array element is more natural, but no strong feelings.

**Review by bh2smith — APPROVED** [review]
Looks great!

File: contracts/BatchExchange.sol (hunk 5 above)
Comment (bh2smith) [reply contracts/BatchExchange.sol:386]: Very good!

File: contracts/BatchExchange.sol (hunk 7 above)
Comment (bh2smith) [reply contracts/BatchExchange.sol]: I'm not sure I understand this comment...

File: test/stablex/batch_exchange.js (hunk 4 above)
Comment (bh2smith) [reply test/stablex/batch_exchange.js]: Yes... It kinda seems the readers interpretation is obfuscated by the definition of the orders more than the point itself. I wonder how we could make this cleaner? Maybe a `deepEquals`  on the order (that shouldn't refer to the values itself, but more the order... Assuming the `orderId` is returned as part or the Order Object, I would think that the values themselves could all be the same.

File: contracts/BatchExchange.sol [hunk 8]
```diff
@@ -401,12 +401,7 @@ contract BatchExchange is EpochTokenLocker {
       * @return encoded bytes representing all orders
       */
     function getEncodedUserOrders(address user) public view returns (bytes memory elements) {
-        for (uint256 i = 0; i < orders[user].length; i++) {
-            elements = elements.concat(
-                encodeAuctionElement(user, getBalance(user, tokenIdToAddressMap(orders[user][i].sellToken)), orders[user][i])
-            );
-        }
-        return elements;
+        return getEncodedUserOrdersPaginated(user, 0, uint16(-1));
```
Comment (bh2smith) [inline contracts/BatchExchange.sol:404]: Does this -1 trick work? I thought this was only a `python` thing!

**Review by bh2smith — APPROVED** [review]
Seems legit!

File: contracts/BatchExchange.sol (hunk 8 above)
Comment (josojo) [reply contracts/BatchExchange.sol:404]: we use it in some places, eg the constructor

**Review by alfetopito — COMMENTED** [review]
:+1:

**Pull request merged**